src/hotspot/share/opto/graphInvariants.cpp

```diff
@@ -0,0 +1,617 @@
+
+    } else if (path.at(path_len - i - 1) == LocalGraphInvariant::OutputStep) {
+      // It's an output
+      ss.print("         -->");
+    } else {
+      ss.print("         ???");
+    }
+    steps.at(path_len - i - 1)->dump("\n", false, &ss);
+  }
+}
+
+struct Pattern : ResourceObj {
+  virtual bool check(const Node* center, Node_List& steps, GrowableArray<int>& path, stringStream&) const = 0;
+};
+
+struct True : Pattern {
+  bool check(const Node*, Node_List&, GrowableArray<int>&, stringStream&) const override {
+    return true;
+  }
+};
+
+struct Bind : Pattern {
+  explicit Bind(const Node*& binding) : _binding(binding) {}
+  bool check(const Node* center, Node_List&, GrowableArray<int>&, stringStream&) const override {
+    _binding = center;
+    return true;
+  }
+  const Node*& _binding;
+};
+
+struct And : Pattern {
+private:
+  template <typename... PP>
+  static void make_helper(And* a, Pattern* pattern, PP... others) {
+    a->_checks.push(pattern);
+    make_helper(a, others...);
+  }
+  static void make_helper(And*) {}
+
+public:
+  template <typename... PP>
+  static And* make(PP... patterns) {
+    And* andd = new And();
+    make_helper(andd, patterns...);
+    return andd;
+  }
+
+  bool check(const Node* center, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    for (int i = 0; i < _checks.length(); ++i) {
+      if (!_checks.at(i)->check(center, steps, path, ss)) {
+        return false;
+      }
+    }
+    return true;
+  }
+  GrowableArray<Pattern*> _checks;
+};
+
+struct HasExactlyNInputs : Pattern {
+  explicit HasExactlyNInputs(uint expect_req) : _expect_req(expect_req) {}
+  bool check(const Node* center, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    if (center->req() != _expect_req) {
+      ss.print_cr("Unexpected number of input. Expected: %d. Found: %d", _expect_req, center->req());
+      for (uint i = 0; i < center->req(); ++i) {
+        Node* in = center->in(i);
+        ss.print("  %d: ", i);
+        if (in == nullptr) {
+          ss.print_cr("nullptr");
+        } else {
+          in->dump("\n", false, &ss);
+        }
+      }
+      return false;
+    }
+    return true;
+  }
+  const uint _expect_req;
+};
+
+struct HasAtLeastNInputs : Pattern {
+  explicit HasAtLeastNInputs(uint expect_req) : _expect_req(expect_req) {}
+  bool check(const Node* center, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    if (center->req() < _expect_req) {
+      ss.print_cr("Too small number of input. Expected: %d. Found: %d", _expect_req, center->req());
+      for (uint i = 0; i < center->req(); ++i) {
+        Node* in = center->in(i);
+        ss.print("  %d: ", i);
+        if (in == nullptr) {
+          ss.print_cr("nullptr");
+        } else {
+          in->dump("\n", false, &ss);
+        }
+      }
+      return false;
+    }
+    return true;
+  }
+  const uint _expect_req;
+};
```

Comment:
Looks like code duplication. Could you make it more general with a callback? You could still create alias structs that already have the correct callback/lambdas.